src/vs/base/common/event.ts

```diff
@@ -732,6 +732,26 @@ export class DebounceEmitter<T> extends PauseableEmitter<T> {
 	}
 }
 
+/**
+ * An emitter which queue all events and then process them at the
+ * end of the event loop.
+ */
+export class MicrotaskEmitter<T> extends Emitter<T> {
+	private _queuedEvents: T[] = [];
+	constructor(options?: EmitterOptions) {
+		super(options);
+	}
+	override fire(event: T): void {
+		this._queuedEvents.push(event);
+		if (this._queuedEvents.length === 1) {
+			queueMicrotask(() => {
+				this._queuedEvents.forEach(e => super.fire(e));
```

Comment:
@lramos15 I think it would be better here to send an array of events at once because then listeners can actually implement in a clever way and not have to react on each event individually. The previous `DebouncedEmitter` is offering a optional `merge` function and we should probably have the same.

Reply:
How would the typings work in this case? Because we have an emitter that emits T. The merge function takes T[] => T. So would T be an array so it would be like Events[] and then Merge would merge an array of arrays? Events[][] => Events[].

Reply:
I wasn't actually suggesting to provide the merge functionality, couldn't the `MicrotaskEmitter` not just always return `Event[]`?

Reply:
I tried accomplishing this but it got quite messy. I think a merge function would be better and then let the user supply an array type and have the merge function concat.

Reply:
Yeah so maybe copying whatever we do for `DebounceEmitter` is fine.

Reply:
Yeah the plane is to change `IEditorChangeEvent` to fire an array and then the merge function will just concat them. This just moves the data structure towards the consumer rather than producer which I think is probably nicer in this case